src/wp-includes/block-template-utils.php

```diff
@@ -1163,14 +1163,40 @@ function get_block_templates( $query = array(), $template_type = 'wp_template' )
 	}
 
 	if ( ! isset( $query['wp_id'] ) ) {
+		$template_files_query = $query;
 		/*
 		 * If the query has found some user templates, those have priority
 		 * over the theme-provided ones, so we skip querying and building them.
 		 */
-		$query['slug__not_in'] = wp_list_pluck( $query_result, 'slug' );
-		$template_files        = _get_block_templates_files( $template_type, $query );
+		$template_files_query['slug__not_in'] = wp_list_pluck( $query_result, 'slug' );
+		/*
+		 * We need to unset the post_type query param because some templates
+		 * would be excluded otherwise, like `page.html` when looking for
+		 * `page` templates. We need all templates so we can exclude duplicates
+		 * from plugin-registered templates.
+		 * See: https://github.com/WordPress/gutenberg/issues/65584
+		 */
+		unset( $template_files_query['post_type'] );
+		$template_files = _get_block_templates_files( $template_type, $template_files_query );
 		foreach ( $template_files as $template_file ) {
-			$query_result[] = _build_block_template_result_from_file( $template_file, $template_type );
+			// The custom templates with no associated post types are available for all post types.
+			if ( isset( $query['post_type'] ) && ! isset( $template_file['postTypes'] ) ) {
```

Comment:
Are these changes not needed in the GB PR?

Reply:
The main difference is that in GB we are hooking into the `get_block_templates` filter, in other words, we are hooking into the _end_ of this function.

In the other hand, in WP core, we can introduce the changes _inside_ the function. While I could have simply called `_get_block_templates_files()` again with `post_type` removed (as it's done in the GB PR), it made more sense to me to reuse the existing call to `_get_block_templates_files()`. It requires some extra logic (basically, the new `if ... else if ...`, but this way we call `_get_block_templates_files()` only once, which I guess is better for performance.

I'm happy to change this approach if you think overcomplicates things.

Reply:
Thanks for explaining. Overall this makes sense to me, though one question: Why is the `isset( $query['post_type'] )` check needed in this line here specifically?

Reply:
It was to avoid entering the `if` clause unnecessarily (as it would go through the `elseif` one, which is more performant). In 5cea08f4d6c186909daf9bfdbbd7828f1c437186 I removed this `isset( $query['post_type'])` and switched the order of the clauses, which hopefully makes the code easier to read.